Assets/MRTK/SDK/Experimental/Elastic/Scripts/QuaternionElasticSystem.cs

```diff
@@ -0,0 +1,154 @@
+            }
+
+            // For clarity and conciseness
+            var k = elasticProperties.HandK;
+            var d = elasticProperties.Drag;
+            var m = elasticProperties.Mass;
+
+            var displacement = Quaternion.Inverse(currentValue) * forcingValue;
+
+            var force = Add(Scale(displacement, elasticProperties.HandK), Scale(currentVelocity, -d));
+
+            var eulers = currentValue.eulerAngles;
+
+            
+            foreach (var interval in extent.SnapPoints )
+            {
+                var nearest = extent.repeatSnapPoints ? GetNearest(eulers, interval) : interval;
+                var nearestQuat = Quaternion.Euler(nearest.x, nearest.y, nearest.z);
+
+                if (Quaternion.Dot(nearestQuat, currentValue) < 0)
+                {
+                    nearestQuat = Scale(nearestQuat, -1.0f);
+                }
+                var snapDisplacement = Quaternion.Inverse(currentValue) * nearestQuat;
+                var snapAngle = Quaternion.Angle(currentValue, nearestQuat);
+
+                var snapFactor = ComputeSnapFactor(snapAngle, extent.SnapRadius);
+
+                force = Add(force, Scale(snapDisplacement, elasticProperties.SnapK * snapFactor));
+            }
+
+            var accel = Scale(force, (1 / m));
+            currentVelocity = Add(currentVelocity, Scale(accel, deltaTime));
+            currentValue = currentValue * Scale(currentVelocity, deltaTime).normalized;
+            currentValue = currentValue.normalized;
+
+            return currentValue;
+        }
+
+        /// <inheritdoc/>
+        public override Quaternion GetCurrentValue() => currentValue;
+
+        /// <inheritdoc/>
+        public override Quaternion GetCurrentVelocity() => currentVelocity;
+
+        private Vector3 GetNearest(Vector3 target, Vector3 interval)
+        {
+            return new Vector3(GetNearest(target.x, interval.x), GetNearest(target.y, interval.y), GetNearest(target.z, interval.z));
+        }
+
+        private float GetNearest(float target, float interval)
+        {
+            return Mathf.Round(target / interval) * interval;
+        }
+
+        private float ComputeSnapFactor(float angleFromPoint, float radius)
+        {
+            // Snap force is calculated by multiplying the "-kx" factor by
+            // a clamped distance factor. This results in an overall
+            // hyperbolic profile to the force imparted by the snap point.
+            return (1.0f - Mathf.Clamp01(Mathf.Abs(angleFromPoint / radius)));
+        }
+
+        private Quaternion Add(Quaternion p, Quaternion q)
+        {
+            return new Quaternion(p.x + q.x, p.y + q.y, p.z + q.z, p.w + q.w);
+        }
+
+        private Quaternion Scale(Quaternion p, float t)
+        {
+            return new Quaternion(p.x * t, p.y * t, p.z * t, p.w * t);
+        }
+    }
+    internal static class QuaternionExtras
```

Comment:
Helper class containing several utility functions for clarifying quaternion operations.